ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/attr/Attr.java

```diff
@@ -0,0 +1,154 @@
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+@UdafDescription(
+    name = "ATTR",
+    description = "The ATTR() aggregation indicates there are multiple values, but only one "
+        + "was expected. For example, if aggregating against a table that semantically should "
+        + "have only one value for a column given a key, this aggregation enables users to "
+        + "indicate that they expect only a single value.\n\n"
+        + "If the aggregation encounters more than a single value for the expected singular "
+        + "column, the entire aggregation will return null."
+)
+public final class Attr {
+
+  private Attr() {
+    // checkstyle complains otherwise
+  }
+
+  @UdafFactory(description = "Collect as a singleton")
+  public static <T> TableUdaf<T, List<Struct>, T> createAttr() {
+    return new Impl<T>();
+  }
+
+  @VisibleForTesting
+  static class Impl<T> implements TableUdaf<T, List<Struct>, T> {
+
+    static final String VALUE = "VALUE";
+    static final String COUNT = "COUNT";
+
+    SqlType inType;
+    Schema entrySchema;
+
+    @Override
+    public void initializeTypeArguments(final List<SqlArgument> args) {
+      this.inType = args.get(0).getSqlTypeOrThrow();
+
+      // we use a list of structs instead of a map here for two reasons:
+      //
+      //   1. our data formats currently only support maps with string keys
+      //   2. null is a valid entry, and most maps don't support null keys
+      //
+      // this should be OK from a complexity perspective because ATTR expects
+      // a unique entry, so only the situations where it is used improperly will
+      // take a runtime hit
+      this.entrySchema = SchemaBuilder.struct()
+          .optional()
+          .field(VALUE, SchemaConverters.sqlToConnectConverter().toConnectSchema(inType))
+          .field(COUNT, Schema.OPTIONAL_INT32_SCHEMA)
+          .build();
+    }
+
+    @Override
+    public Optional<SqlType> getAggregateSqlType() {
+      return Optional.of(SqlTypes.array(
+          SchemaConverters.connectToSqlConverter().toSqlType(entrySchema)
+      ));
+    }
+
+    @Override
+    public Optional<SqlType> getReturnSqlType() {
+      return Optional.of(inType);
+    }
+
+    @Override
+    public List<Struct> initialize() {
+      return new ArrayList<>();
+    }
+
+    @Override
+    public List<Struct> aggregate(final T current, final List<Struct> agg) {
+      final List<Struct> out = new ArrayList<>(agg);
+      update(out, current, 1);
+      return out;
+    }
+
+    @Override
+    public List<Struct> merge(final List<Struct> one, final List<Struct> two) {
+      // use O(n^2) algorithm here because in practice each of these lists
```

Comment:
Based on the other discussions on this thread, I'd note that the core of this class is likely how we'd want to implement Collect_Set as a TableUDAF.

Given that, it may be amusing to think through reducing the complexity.  That said, I'm totally happy to see that deferred to another ticket.  I think we should be happy with the types as-is.  (And changing those is trickier.)

In the general case, would it be faster to sort each list and merge them together?  Naively, seems like that's `O(n log n) + O(m log m) + O(n + m)` rather than `O(n*m)`.

Reply:
Now that I'm giving this some more thought.... if we use insertion sort, this may actually amortize to `O(n)` sort because we know that we're starting with a sorted list every time. Let me noodle with that a bit in the IDE and see if my intuition is correct.

Reply:
erm... just realized I'd need to implement the comparator for complex values as well, which would be rather difficult...